### ISSUE TITLE
addIceCandidate may not need to throw an error when no remoteDescription

### ISSUE THREAD

**Issue by Sean-Der** [w3c/webrtc-pc#2519]
(This feedback is from [Pion](https://github.com/pion/webrtc) not my full-time job, go ahead and delete if that isn't ok)

This is defined [here](https://github.com/w3c/webrtc-pc/blob/master/webrtc.html#L3032-L3035)

```
If remoteDescription is null return a promise rejected with a newly created InvalidStateError.
```
----

I don't believe this restriction is needed, and many popular WebRTC libraries just circumvent this check. In [simple-peer](https://github.com/feross/simple-peer/blob/master/index.js#L209-L211) the library caches candidates, and then adds them.

I do the same in Pion.

I can get more datapoints from other implementations if that is helpful. I have seen this pattern in many private code bases as well.

----

From an implementer standpoint this is really easy to handle. If you get an ICE candidate before you have ICE details just cache them, and when you have enough details you can start connectivity checks.

I work with a lot of people learning WebRTC/building things for the first time because of Pion. This error is constant, and is probably my #2 issue (after H264 in Chrome questions/not support all profiles). Especially with the new rush of interest with COVID this is something that I will sometimes have people sharing logs multiple times in Slack a day with this error.

**Comment by fippo** [comment]
adding a candidate before the remote description is a programming error. Throwing it *forces* people to fix their programming and/or signaling protocol errors.

**Comment by Sean-Der** [comment]
It is an error only because this document defines it as such, if the document changes it isn't an error anymore. Developers have enough problems to worry about, why make them suffer from a death by a thousands cuts on stuff like this?

----

With this we require
* Either people put logic in their Client code to cache candidates
* Make their signaling servers smarter, making them cache ice candidates. This prevents people from using simple pub/sub code.

If I am a new developer and I open the [firebase example](https://github.com/webrtc/FirebaseRTC/blob/solution/public/app.js#L173-L183) I see that we have this event handler after `setRemoteDescription`, why not just move it before? There is nothing explicit about this.  

----

If we removed this assert and moved the 'candidate caching' into the Clients all this would go away. This restriction also may not show up during local development, then when they go to prod they are stuck debugging this.

**Comment by alvestrand** [comment]
This happens in practice when the path for signaling candidates and the path for signaling state updates are uncoordinated, so that a remote candidate may race a remote description.

Stats show that rejection because of no remote description happens in 0.197% of AddIceCandidates calls. It's by far the most common reason for failed AddIceCandidates calls.

**Comment by fippo** [comment]
> Stats show that rejection because of no remote description happens in 0.197% of AddIceCandidates calls

You're looking at global metrics which contain a lot of noise. I've seen error rates below 0.023 % (or five sigma) for a particular system. That was after significant efforts to analyse and eliminate other error causes.

**Comment by Sean-Der** [comment]
@fippo Is that stat from a system you controlled? The only metric that matters here to me is the global one. 

When people realize they have this error they will go fix it.

**Comment by jan-ivar** [comment]
Hi @Sean-Der I worry this is covering up a bug somewhere. Does your signaling guarantee order?

WebRTC relies on a serialized signaling channel, so queuing ICE candidates is an [anti-pattern](https://github.com/web-platform-tests/wpt/issues/19866). It
 1. complicates negotiation & renegotiation
 2. covers up signs of browser and/or application bugs
 3. defeats Trickle ICE in [some cases](https://stackoverflow.com/a/36851163), slowing down connection

The spec says `onicecandidate` [cannot fire](https://w3c.github.io/webrtc-pc/#ref-for-dfn-release-early-candidates-1) before SLD, guaranteeing this signaling order:
```
offer candidate candidate candidate
```
and in the reciprocal direction
```
answer candidate candidate candidate
```

This becomes important during rapid re-negotiation and ICE restarts, where we might see:
```
offer candidate candidate offer candidate candidate candidate
```
Does the second candidate belong to the first or second offer? Also, most queuing I've seen only works on initial negotiation, which may lead to missed candidates and intermittents if it was wallpapering a real timing hole in a browser or application.

Negotiation being an async state machine, strict state management is one of the few guide-rails we offer, so I don't think we should relax them.

**Comment by fippo** [comment]
@Sean-Der global stats have proven to be :hankey: in the past, e.g. when chrome rolled out AEC3 (which ended in a desaster on high-price headsets) or in the failed first attempt to deprecate DTLS 1.0.

**Comment by Sean-Der** [comment]
@jan-ivar This isn't a bug I am having personally. I have an Open Source project [Pion](https://github.com/pion/webrtc) and I see lots of developers struggling with this. Many of the people coming to the project are using WebRTC for the first time.

> Does your signaling guarantee order?
This is going to bite people. I don't think I have seen a WebRTC tutorial that calls this out explicitly. I think we are asking too much of new developers to deal with this.

> covers up signs of browser and/or application bugs (go ★ it)
IMO not a bug. You are making WebRTC implementations more complicated with this requirement. If the ICE agent can just emit candidates when ready things would be better. You are now forcing more logic for the agent to be aware of the PeerConnection state.

> defeats Trickle ICE in some cases, slowing down connection
From reading the issue it sounds like this assert is the actual thing causing the issue. If we didn't have this requirement people could emit candidates at any time. The slowing down of trickle ICE (and extra logic) has happened because of the SetRemoteDescription requirement.

> most queuing I've seen only works on initial negotiation
This is all I am looking for! I think after the initial negotiation all the logic stays the same.

-----

I would just love to see this sharp edge removed (and I am still not convinced it is going to cost anything). If every user has to go look up this detail that is a lot of man hours wasted.

**Comment by jan-ivar** [comment]
@Sean-Der Please show me a simple example demonstrating the problem.

> > most queuing I've seen only works on initial negotiation
> 
> This is all I am looking for! I think after the initial negotiation all the logic stays the same.

The same methods and logic are used on renegotiation, so we have to think about that as well. E.g.
* App adding new tracks for a participant or screen-sharing sends additional ICE candidates
* App defers adding tracks until after initial negotiation
* App restarts ICE e.g. after connection failure
* App stops a BUNDLE transceiver

**Comment by jan-ivar** [comment]
> > Does your signaling guarantee order?
> 
> This is going to bite people. I don't think I have seen a WebRTC tutorial that calls this out explicitly. I think we are asking too much of new developers to deal with this.

What signaling channel are people bitten by this using?

**Comment by fippo** [comment]
> This is going to bite people.

Ordering is a pretty natural guarantee actually. See https://tools.ietf.org/html/rfc6120#section-10.1 if you want a more formal description. It is a fundamental assumption in protocol design. For example it ensures that the join confirmation for a chatroom arrives before you get any messages. E.g. you get
```
C: join X
S: you joined X
S: here is an offer from the room jingle focus
```
I've only once seen a signaling protocol which didn't have this property. Signaling protocols aren't simple and there is a point where one should strongly consider whether reinventing the wheel makes sense. I'm biased of course and still prefer jingle over sending SDP on the wire :-)

Suppose you get candidates with a ufrag X. Then you get a remote description with ufrag Y. Do you reject the remote description or silently discard the generation with ufrag X?

**Comment by Sean-Der** [comment]
> @Sean-Der Please show me a simple example demonstrating the problem.
>

Something like [this](https://jsfiddle.net/3y8zebL9/)? I took this from a highly voted answer on SO [here](https://stackoverflow.com/a/49055883/5472819). I add the `setTimeOut` just to make it happen everytime.

If I am a new developer reading the official [webrtc.org docs on Trickle ICE](https://webrtc.org/getting-started/peer-connections) why would I set onicecandidate after `SetLocalDescription`?

> What signaling channel are people bitten by this using?
Disregard this, it is really a niche case. If we discuss this it is distracting from the main discussion IMO. Sorry for bringing it up

> The same methods and logic are used on renegotiation, so we have to think about that as well. E.g.

I don't think this adds any new problems for renegotation? After the first remote-description has been received this remote candidate cache is never used again.

**Comment by fippo** [comment]
> Something like this?

That is a known problem but as far as I know it only affects local examples. See https://bugs.chromium.org/p/chromium/issues/detail?id=960718
The tl;dr is that you call SLD and therefore start generating candidates before SRD is called.
With setTimeout you are intentionally screwing up the order. If it happens without the setTimeout... please file a bug!

> If I am a new developer reading the official webrtc.org docs

Please note that webrtc.org is not "official". In particular the codelab there contains several errors and antipatterns and clearly lacks review.

**Comment by jan-ivar** [comment]
> Something like this?

That example was correctly written until you added a 1 second delay after SLD. Again, from [SO](https://stackoverflow.com/a/36851163/918910):

> 1. Send pc.localDescription no later than in the setLocalDescription success callback.
> 2. Expect pc.onicecandidate to start firing immediately after that callback. Send them. 

Signaling is implicit & immediate in local-loop demos, and you never "sent" your offer (by not calling answerer.SRD(offer)) before your 1 second delay, while still sending candidates without delay, thus breaking order. E.g. it's as if you'd signaled:
```
candidate candidate candidate (1 second pause) offer
```
Note it still connects, since local candidates get inserted into the local offer as well, so this is just error spam from unnecessary trickling.

IOW, if you're going to delay initial connection by 1 second, do it anywhere else, since Trickle ICE exists solely to save that second.

> > I add the `setTimeOut` just to make it happen everytime.
> 
> If it happens without the setTimeout... please file a bug!

Yes please, and test multiple browsers. Without your `setTimeout`, even the local-loop case should be air tight, thanks to peer connection's [operations chain](https://w3c.github.io/webrtc-pc/#dfn-operations-chain) (once SRD() is on that chain, addIceCandidate cannot race it).

**Comment by Sean-Der** [comment]
> if you're going to delay initial connection by 1 second, do it anywhere else
People are going to abuse the API, most times accidentally. I am just trying to demonstrate where the API is needlessly fragile.

This feels like 'Undefined Behavior' conversations. You can tell developers `Do not do that`, but they are going to do it anyway. Why not design the API to be robust. You can chase help people on SO every time this comes up, but why not fix the root cause?

-----

I think we will just have to agree to disagree. Since I don't have W3C membership the conversations ends here, I appreciate you taking your time to discuss it though! Thank you for all your hard work on the webrtc-pc API :)

**Comment by jan-ivar** [comment]
We've been working over the last year to [make the API robust](https://webrtchacks.com/min-duration-series-part-1-perfect-negotiation/). I recommend the [spec example](https://w3c.github.io/webrtc-pc/#perfect-negotiation-example).

> why not fix the root cause?

Because there's no guarantee that allowing JS to do things out of order on initial negotiation won't then happen again on renegotiation, as I mentioned https://github.com/w3c/webrtc-pc/issues/2519#issuecomment-622085409. Most likely it would, and then it's much less clear how we'd resolve it correctly, as mentioned in https://github.com/w3c/webrtc-pc/issues/2519#issuecomment-622272086. I'm happy to pick up the conversation there if you have ideas on how to solve it.

I also suspect the "root cause" is or may have been a bug in Chrome. I'm going to rip out some wpt helpers to help find it, since being too lenient in tests appears to be coming back to bite us.

**Comment by fippo** [comment]
> This feels like 'Undefined Behavior' conversations. You can tell developers Do not do that, but they are going to do it anyway. Why not design the API to be robust

This is a philosophical discussion and one which has value. See [Postel was wrong](https://tools.ietf.org/html/draft-iab-protocol-maintenance-04) for one opinion. I'm on the "throw errors, dont assume the browser implementors are smarter" side.
Doing a timing test for addIceCandidate that reliably covers all native code parts in *all* browsers is going to be somewhat impossible.

Now where did I put the document with "p2p ice failure rates dropped *considerably* when postponing addIceCandidate slightly in chrome...
(btw, you can still raise issues even without being a w3c member or invited expert... so far @dontcallmedom hasn't tried to stop me :-))

**Comment by jan-ivar** [comment]
> where did I put the document with "p2p ice failure rates dropped considerably when postponing addIceCandidate slightly in chrome...

@fippo If you have similar data for Firefox that would help show if this is something other than an implementation bug.

**Comment by jan-ivar** [comment]
From editor's meeting: Agreement that throwing an error is the right thing to do.

We think the root cause was fixed in https://bugs.chromium.org/p/chromium/issues/detail?id=1019222

**Comment by Sean-Der** [comment]
@jan-ivar This came up again. A user was interfacing with an SFU (not using libwebrtc at all)

I think that adds another layer to the conversation! All the conversations before are in the 'single threaded' world of the browser. What about servers that don't even implement the W3C API?

**Comment by aldovincenti** [comment]
I tried the [perfect negotiation example](https://w3c.github.io/webrtc-pc/#perfect-negotiation-example) with a simple signaling server (Node.js + Socket.io, WebSocket protocol), and unfortunately, I found that sometimes I receive ICE candidates before the remote description. I had to queue ICE candidates manually, and only for this limitation. I think this problem is more common than we suppose, not surprisingly, as @Sean-Der mentioned, many libraries implement a queue. Please reconsider this limitation, it makes implementing WebRTC difficult and frustrating.

**Comment by jech** [comment]
Suppose you delegate caching of ICE candidates to the library.  Now what happens if you do:

  * pc.setLocalDecription
  * pc.setRemoteDescription
  * pc.addICECandidate
  * pc.restartICE

How would the library know whether the ICE candidate belongs to the first ICE session (and therefore be applied immediately) or to the second ICE session (and therefore be cached)?

**Comment by fippo** [comment]
> How would the library know whether the ICE candidate belongs to the first ICE session

That is why the ufrag was added at some point.

**Comment by tran-simon** [comment]
> Signaling is implicit & immediate in local-loop demos, and you never "sent" your offer (by not calling answerer.SRD(offer)) before your 1 second delay, while still sending candidates without delay, thus breaking order. E.g. it's as if you'd signaled:

What if my signaling channel is slow and/or unreliable? I'm sometimes getting the error that I'm adding ice candidates before `setRemoteDescription` is called, because I'm using a more complex signaling channel that is unreliable (some messages take longer to reach the receiver than others).

The only way I can reproduce the problem consistently is by adding a delay before `setRemoteDescription`. This helps mimic real world condition, where ice candidate messages might reach the receiver before the WebRTC offer.

**Comment by alvestrand** [comment]
The simplest client-side fix (which the WPT code had implemented at one point, but it got rolled out because of the still somewhat contentious claim that this can never happen without channel reordering) is to accumulate candidates until the signaling state returns to stable (after SLD/SRD), and then accept them all.

**Comment by Sean-Der** [comment]
@alvestrand Would you recommend that all WebRTC developers implement this logic? 

Working with new WebRTC users this is one of the most common questions I get. I also don't understand why this was resolved with a link to a Chromium bug? 

This is something that in applicable to any implementer.

**Comment by fippo** [comment]
I haven't seen this much in the last three years but by coincidence this came up on [stackoverflow this week](https://stackoverflow.com/questions/76130700/javascript-callback-would-wait-async-function-to-complete) and with an example that explains why this can happen sometimes.
The flow that breaks is shown by this fiddle:
https://jsfiddle.net/fippo/b4hLpc2d/2/

The root cause here is the messageHandler is not an async function and gets called while there is still a call in progress. That is a problem of the signaling and not a WebRTC problem.
The second problem is calling getUserMedia before calling setRemoteDescription. Changing this to call setRemoteDescription before getUserMedia makes things work just fine as show by this variant of the fiddle:
https://jsfiddle.net/fippo/b4hLpc2d/3/

If you want to cache candidates before SRD I wish you good luck with specifying all the edge cases...

**Comment by Sean-Der** [comment]
@fippo Could it be added to the spec that WebRTC only supports ordered/lossless signaling?

I personally would prefer if WebRTC APIs were made more robust. I still don’t understand why we can’t ‘upstream the boilerplate’ that WebRTC libraries are already doing.

**Comment by aboba** [comment]
@sean-der Signaling is out of scope of the WebRTC API. So an application could signal using [RFC 1149](https://en.wikipedia.org/wiki/IP_over_Avian_Carriers#:~:text=In%20computer%20networking%2C%20IP%20over%20Avian%20Carriers%20%28,D.%20Waitzman%2C%20and%20released%20on%20April%201%2C%201990.).

**Comment by jech** [comment]
> @Sean-Der Signaling is out of scope of the WebRTC API. So an application could signal using RFC 1149.

RFC 1149 is a network layer protocol.  You'd use a transport layer protocol between RFC 1149 and signalling, and I think what Sean meant is that he'd like to simplify the implementation of signalling over X over RFC 1149 in the specific case where X preserves message ordering.

@fippo is claiming that the simplification suggested by Sean will not solve all issues in the case where X does not preserve ordering.  He's probably right (I frankly don't know), but I don't feel that's a valid argument against simplifying the implementer's job in the specific case where X is ordered.

**Comment by lgrahl** [comment]
If I understand the issue correctly, there is no ordering violation on the signalling layer but rather an API call ordering violation. Because SLD and SRD are async, candidates will have to be held back even though they may have already been received.

IMO: I don't see how the suggestion (to accept ICE candidates regardless of the state) would make the API more robust. I think it would provide a false sense of simplicity and hide usage errors. That is pretty much the opposite of what I'd want as a developer. The peer connection is already too much of a black box when it comes to its internal state (e.g. ICE dropping candidate pairs), I don't think we should make it even less opaque. Moreover: If we disregard state here, what about other races on the peer connection state machine that are just less apparent? Would we have to address those as well and where do we draw the line?

**Comment by jech** [comment]
> Can you give examples of standardized protocols that do not preserve message order?

HTTP.

**Comment by Sean-Der** [comment]
@fippo 

The implementation I would like is caching candidates by the `ufrag`

* If no `ufrag` is specified with the candidate throw an error
* If a `ufrag` is specified cache the candidate
* When a `SRD` happens with the matching `ufrag` use the cached candidates

----

I don't understand this point

```
If you think that your signaling protocol doesn't require order, how do you deal with a subsequent offer with 5 m-lines being processed ahead of the original one with 2 m-lines? That may work but applying the two m-line offer will fail since it looks (from an API PoV) like you are removing m-lines which is a no-no.
```

This is covered by `Session Version` in `o` value (I believe)?

**Comment by fbarbat** [comment]
Hey, just sharing here our experience.
I hit this issue while trying to integrate webRTC to our app. We were already using Firestore so it was handy to use it too for the signaling. Specifically, clients write messages (documents) to specific subcollections owned by individual users and it works fine most of the time. But Firestore doesn't guarantee order of insertion for the listeners, I could reproduce out of order messages by causing problems to the network.
Having WebRTC dealing with out of order messages would have made the integration path easier and safer. We are going to try @Sean-Der approach and track errors and out of order messages. If we see problems, we might need to make more drastic changes to our architecture... But if WebRTC by itself would handle that and avoid us some work, that would be amazing.
Is there any problems in queuing ICE candidates by ufrag and using them after processing the corresponding SDP? Does the order between candidates matter?
